src/juju_doctor/probes.py

```diff
@@ -0,0 +1,53 @@
+"""Helper module to wrap and execute probes."""
+
+import importlib.util
+import inspect
+import logging
+from dataclasses import dataclass
+from pathlib import Path
+
+from rich.logging import RichHandler
+
+from juju_doctor.artifacts import Artifacts
+
+SUPPORTED_PROBE_TYPES = ["status", "bundle", "show_unit"]
+
+logging.basicConfig(level=logging.WARN, handlers=[RichHandler()])
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class Probe:
+    """A probe that can be executed via juju-doctor."""
+
+    name: str
+    uri: str
+    original_path: Path  # path in the source folder
+    path: Path  # path in the temporary folder
+
+    def run(self, artifacts: Artifacts):
+        """Run the probe."""
+        module_name = "probe"
+        spec = importlib.util.spec_from_file_location(module_name, self.path.resolve())
+        if not spec:
+            raise ValueError(f"Probe not found at its 'path': {self}")
+        module = importlib.util.module_from_spec(spec)
+        if spec.loader:
+            spec.loader.exec_module(module)
+        functions = inspect.getmembers(module, inspect.isfunction)
+        for name, func in functions:
+            match name:
```

Comment:
- Structural pattern matching seems like an overkill for the current use. Would you be able to combine `artifacts.*` in the match expression to avoid needing the if-statements inside?
- Would need to bump to [3.10](https://peps.python.org/pep-0636/) https://github.com/canonical/juju-doctor/blob/9d102ae7d93456d637c399af80d48ebee0bd355d/pyproject.toml#L14

Reply:
That logic has to go somewhere:
- we could make a dictionary with `status`, `bundle` etc. as top-level keys, but even then we'd have to check if the keys exist
- if we don't raise "manually", we can't tell the user "no artifact has been provided", and the probe will fail with a possibly confusing error message
- we could move this logic into `artifacts.py`, but I don't really see the point

Reply:
Python version has been bumped